### PR TITLE
PE 1719/ Share Page Activity & Details Panel

### PR DESCRIPTION
**Pull request by javdhu**
Redesigns file share page to include activity and details panel.

### CODE REVIEW

**Review by karlprieb — COMMENTED** [review]
I think we can have, at least, some drafts for tests and a tech debt ticket to implement them next week when we will be able to run tests again.

When we call `getAllFileEntitiesWithId` we could get the `owner` of the first tx and check if the owner is the same when we call `getLatestFileEntityWithId`. So we ensure that the file shared is from the same wallet that first uploaded it.

**Review by karlprieb — COMMENTED** [review]
can we check owner on `getLatestFileEntityWithId` as well?
Also do we have tests for `getLatestFileEntityWithId`?

File: test/services/arweave/arweave_service_test.dart [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+const gatewayUrl = 'https://arweave.net';
+void main(List<String> args) {
+  group('Arweave Service Tests', () {
+    final arweave = ArweaveService(
+      Arweave(
+        gatewayUrl: Uri.parse(gatewayUrl),
+      ),
+    );
+    test('AllFileEntitiesWithId returns all the file entities for a known Id',
```
Comment (karlprieb) [inline test/services/arweave/arweave_service_test.dart:13]: add a test for when `getAllFileEntitiesWithId` returns null
why did you add `expect(true, isTrue);`?

File: lib/models/file_revision.dart [hunk 2]
```diff
@@ -35,8 +35,9 @@ extension FileEntityExtensions on FileEntity {
   /// Converts the entity to an instance of [FileRevisionsCompanion].
   ///
   /// This requires a `performedAction` to be specified.
-  FileRevisionsCompanion toRevisionCompanion(
-          {required String performedAction}) =>
+  FileRevisionsCompanion toRevisionCompanion({
+    required String performedAction,
```
Comment (karlprieb) [inline lib/models/file_revision.dart:39]: I know that's not the scope of this ticket, but can we have an enum here instead of string?

File: test/services/arweave/arweave_service_test.dart [hunk 3]
```diff
@@ -0,0 +1,31 @@
+import 'dart:math';
+
+import 'package:ardrive/services/services.dart';
+import 'package:arweave/arweave.dart';
+import 'package:flutter_test/flutter_test.dart';
+
+const gatewayUrl = 'https://arweave.net';
+void main(List<String> args) {
+  group('Arweave Service Tests', () {
```
Comment (karlprieb) [inline test/services/arweave/arweave_service_test.dart:9]: this test is not reliable... it depends on network to work and will break in the future.
We should mock http calls here with `MockClient` from `http/testing`.

File: lib/models/file_revision.dart (hunk 2 above)
Comment (matibat) [reply lib/models/file_revision.dart:39]: You can also simply use the getter `MyEnum.EnumValue.name` which does the same without having to split it. See: [name property](https://api.flutter.dev/flutter/dart-core/EnumName/name.html).

File: test/services/arweave/arweave_service_test.dart [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    );
+    test('AllFileEntitiesWithId returns all the file entities for a known Id',
+        () async {
+      const knownFileId = '0f029ecf-6593-4942-ad07-a57401e3861c';
+      const knownRevisionCount = 9;
+      final fileEntities = await arweave.getAllFileEntitiesWithId(knownFileId);
+      expect(fileEntities?.length, equals(knownRevisionCount));
+    });
+    test('AllFileEntitiesWithId returns null for invalid Id', () async {
+      final invalidFileId = Random().nextInt(1000).toString();
```
Comment (matibat) [inline test/services/arweave/arweave_service_test.dart:23]: Let's use an explicit value instead of making use of Random to make the test deterministic

File: test/services/arweave/arweave_service_test.dart (hunk 3 above)
Comment (matibat) [reply test/services/arweave/arweave_service_test.dart:9]: We are currently injecting an `Arweave` intense into `ArweaveService`, if we also injected the `ArtemisClient` we could mock both

**Review by matibat — COMMENTED** [review]
Left some comments

File: lib/models/file_revision.dart (hunk 2 above)
Comment (karlprieb) [reply lib/models/file_revision.dart:39]: Can you create a tech debt for it?

File: lib/pages/shared_file/shared_file_page.dart [hunk 5]
```diff
@@ -17,91 +20,142 @@ class SharedFilePage extends StatelessWidget {
...
+                              label: Text(appLocalizationsOf(context).download),
+                              onPressed: () => promptToDownloadSharedFile(
+                                context: context,
+                                fileId: state.fileRevisions.last.fileId,
+                                fileKey: state.fileKey,
+                              ),
+                            ),
+                            const SizedBox(height: 16),
+                            _buildReturnToAppLink(context),
+                          } else if (state is SharedFileNotFound) ...{
```
Comment (matibat) [inline lib/pages/shared_file/shared_file_page.dart:100]: What does the "..." do in here? `if() ...{}`
Is it destructuring?

File: lib/services/arweave/arweave_service.dart [hunk 6]
```diff
@@ -533,6 +535,66 @@ class ArweaveService {
...
+        FirstFileEntityWithIdOwnerQuery(
+            variables: FirstFileEntityWithIdOwnerArguments(fileId: fileId)));
+
+    if (firstOwnerQuery.data!.transactions.edges.isEmpty) {
+      return null;
+    }
+
+    final fileOwner =
+        firstOwnerQuery.data!.transactions.edges.first.node.owner.address;
+    while (true) {
```
Comment (matibat) [inline lib/services/arweave/arweave_service.dart:554]: I always fear infinite loops. Could we track a concise exit condition instead of using break?
Such as...
```ts
var hasNextPage = true;
while(hasNextPage) {
  ...
  hasNextPage = !queryEdges.isEmpty && pageInfo.hasNextPage;
```

Comment (matibat) [reply lib/services/arweave/arweave_service.dart:554]: Or maybe a `do {} while ()` 🤔

File: test/services/arweave/arweave_service_test.dart (hunk 3 above)
Comment (matibat) [reply test/services/arweave/arweave_service_test.dart:9]: Perfect! Now the class is testable!
We're just missing the update in the tests to remove any call to the network

**Review by matibat — COMMENTED** [review]
Left some comments - _TODO_ I still want to manually test it

File: lib/blocs/shared_file/shared_file_cubit.dart [hunk 7]
```diff
@@ -56,20 +60,73 @@ class SharedFileCubit extends Cubit<SharedFileState> {
...
+  Future<List<FileRevision>> computeRevisionsFromEntities(
+    List<FileEntity> fileEntities,
+  ) async {
+    late FileRevision oldestRevision;
+
+    fileEntities.sort((a, b) => a.createdAt.compareTo(b.createdAt));
+    oldestRevision = fileEntities.first.toRevision(
+      performedAction: RevisionAction.create,
+    );
+    fileEntities.removeAt(0);
```
Comment (thiagocarvalhodev) [inline lib/blocs/shared_file/shared_file_cubit.dart:87]: Why do we are removing the first index? Consider create a private function or adding a comment to explain that.

File: lib/blocs/shared_file/shared_file_cubit.dart [hunk 8]
```diff
@@ -56,20 +60,73 @@ class SharedFileCubit extends Cubit<SharedFileState> {
...
-    } else {
-      final file = await _arweave.getLatestFileEntityWithId(fileId, fileKey);
+      return;
+    }
+    final allEntities = await _arweave.getAllFileEntitiesWithId(
+      fileId,
+      fileKey,
+    );
+    if (allEntities != null) {
+      {
```
Comment (thiagocarvalhodev) [inline lib/blocs/shared_file/shared_file_cubit.dart:122]: Remove this extra `{`

File: test/test_utils/fake_data.dart [hunk 9]
```diff
@@ -1,19 +1,19 @@
...
+
+// Entity Sample Data
+
+const testEntityName = 'test-name';
+const testEntitySize = 10;
+const testEntityDataContentType = 'test-data-mime-type';
+const testEntityDataTxId = 'FxgiSM3JOAsjnZF-Yzv2z7h-Io4K6GmCarhQ0qAUL6I';
+final testEntityId = const Uuid().v4();
+final driveId = const Uuid().v4();
+final rootFolderId = const Uuid().v4();
```
Comment (thiagocarvalhodev) [inline test/test_utils/fake_data.dart:19]: Why add `final` to a `const`?

Comment (thiagocarvalhodev) [reply test/test_utils/fake_data.dart:19]: OH, got it!

**Review by karlprieb — DISMISSED** [review]

File: test/test_utils/fake_data.dart (hunk 9 above)
Comment (matibat) [inline test/test_utils/fake_data.dart:19]: These inputs are still not deterministic but random x.x
Let's try to use an explicit value for each

File: test/services/arweave/arweave_service_test.dart (hunk 4 above)
Comment (matibat) [reply test/services/arweave/arweave_service_test.dart]: The inputs are still random values - see my review comment

**Review by matibat — COMMENTED** [review]
Thanks for addressing my previous comments. Left a comment regarding the tests

**Review by karlprieb — APPROVED** [review]
